psh/ping/ping.c

```diff
@@ -134,28 +136,35 @@ static int ping_echo(int fd, char *data, int len)
 }
 
 
-static int ping_reply(int fd, char *data, size_t len, char *addrstr, int addrlen)
+static int ping_reply(int fd, uint8_t *data, size_t len, char *addrstr, int addrlen)
 {
 	struct sockaddr_in rsin;
 	socklen_t rlen = sizeof(rsin);
 	struct icmphdr *icmphdr;
 	int bytes;
 	uint16_t chksum;
 
-	if ((bytes = recvfrom(fd, data, len, 0, (struct sockaddr *)&rsin, &rlen)) <= 0) {
-		if (errno == EAGAIN || errno == EWOULDBLOCK)
-			fprintf(stderr, "Host timeout\n");
-		else
-			fprintf(stderr, "ping: Fail to receive packet on socket!\n");
-		return -1;
-	}
+	do {
+		if ((bytes = recvfrom(fd, data, len, 0, (struct sockaddr *)&rsin, &rlen)) <= 0) {
+			if (errno == EAGAIN || errno == EWOULDBLOCK)
+				fprintf(stderr, "Host timeout\n");
+			else
+				fprintf(stderr, "ping: Fail to receive packet on socket!\n");
+			return -1;
+		}
 
-	if (bytes < sizeof(struct iphdr) + sizeof(struct icmphdr)) {
-		fprintf(stderr, "ping: Received msg too short (%d)!\n", bytes);
-		return -1;
-	}
+		if (bytes < (int)(sizeof(struct iphdr) + sizeof(struct icmphdr))) {
+			fprintf(stderr, "ping: Received msg too short (%d)!\n", bytes);
+			return -1;
+		}
 
-	bytes -= sizeof(struct iphdr);
+		icmphdr = (struct icmphdr *)(data + sizeof(struct iphdr));
+		bytes -= sizeof(struct iphdr);
+
+		if (icmphdr->un.echo.id != ping_common.myid) {
```

Comment:
Isn't `ntohs` needed here?

Reply:
![image](https://github.com/phoenix-rtos/phoenix-rtos-utils/assets/3939983/0ca87c9b-6d5f-45b7-8a6c-81c6392e58f5)

maybe stupid optimization (still, busybox does the same)